pkg/resource/open_id_connect_provider/hooks.go

```diff
@@ -0,0 +1,301 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package open_id_connect_provider
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strings"
+
+	ackcompare "github.com/aws-controllers-k8s/runtime/pkg/compare"
+	ackcondition "github.com/aws-controllers-k8s/runtime/pkg/condition"
+	ackerr "github.com/aws-controllers-k8s/runtime/pkg/errors"
+	ackrtlog "github.com/aws-controllers-k8s/runtime/pkg/runtime/log"
+	svcsdk "github.com/aws/aws-sdk-go/service/iam"
+	corev1 "k8s.io/api/core/v1"
+
+	svcapitypes "github.com/aws-controllers-k8s/iam-controller/apis/v1alpha1"
+	commonutil "github.com/aws-controllers-k8s/iam-controller/pkg/util"
+)
+
+func (rm *resourceManager) customUpdateOpenIDConnectProvider(
+	ctx context.Context,
+	desired *resource,
+	latest *resource,
+	delta *ackcompare.Delta,
+) (updated *resource, err error) {
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.sdkUpdate")
+	defer func() {
+		exit(err)
+	}()
+
+	if immutableFieldChanges := rm.getImmutableFieldChanges(delta); len(immutableFieldChanges) > 0 {
+		msg := fmt.Sprintf("Immutable Spec fields have been modified: %s", strings.Join(immutableFieldChanges, ","))
+		return nil, ackerr.NewTerminalError(fmt.Errorf(msg))
+	}
```

Comment:
I see you got this from the code-generator. Honestly we should really be generating this stub even though no update implementation was found. Another error that I found in it, actually, is that we are meant to treat immutable changes as `Advisory` conditions (not `Terminal`). See https://github.com/aws-controllers-k8s/runtime/blob/cb8a3ab18493003e681fdaf4b45adcfdf4229613/apis/core/v1alpha1/conditions.go#L44-L47

Reply:
Yup ideally this should be injected in the main update function. We can open an issue and follow up with this.

Reply:
What do you propose as the correct course of action here @RedbackThomson ? 

A) Postpone this PR until `code-generator` is updated to produce this automatically, or instead B) modify this to apply an `Advisory` condition instead of `Terminal`?

If B), then I don't see anything currently in the `runtime` [here](https://github.com/aws-controllers-k8s/runtime/blob/35dd8c179dec95abeec3e84c0694cf6b9e8a275c/pkg/condition/condition.go#L27) that could be used to set such a condition, so perhaps that would necessitate making an update to `runtime`?

Please advise!

Reply:
I have changed the behavior here such that an attempt to modify an immutable field in this update path will result in an Advisory condition being applied.  Note that this may result in the ACK resource's Spec being out of sync with the upstream AWS resource (specifically, in this case, w/r/t to the `URL` field) even though `ResourceSynced` is set to `True`.  Is that acceptable?

Reply:
Hmm yes `Advisory` and `ResourceSynced` can only be mutually exclusive since `ResourceSynced` indicates that our AWS resource matches our spec. The `EnsureConditions` code in the common runtime should be place to ensure this behaviour. Perhaps for now we can leave `Advisory` out of this PR and come back to this when we can handle this for the general case.

Reply:
Just undid my last commit, behavior now as originally written -- modifying immutable field results in a `TerminalError`